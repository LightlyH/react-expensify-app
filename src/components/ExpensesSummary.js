import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from './../selectors/expenses';
import selectExpensesTotal from './../selectors/expenses-total'; 


export const ExpensesSummary = props => {
    const expenseWord = props.expenses.length === 1 ? 'expense' : 'expenses'; 

    return (
        <div>
            <h1>Viewing {props.expenses.length} {expenseWord} totaling {numeral(props.expensesTotal / 100).format('$0,0.00')}</h1>
        </div>
    );
}; 
 
const mapStateToProps = state => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);

    return {
        expenses: visibleExpenses,
        expensesTotal: selectExpensesTotal(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpensesSummary);

// this alternative also works:
// export const ExpensesSummary = ({ expenseCount, expensesTotal }) => { // destructuring prop
//     const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
//     const formattedExpensesTotal = numeral(expensesTotal).format('$0,0.00');

//     return (
//         <div>
//             <h1>Viewing {expenseCount} {expenseWord} totaling {formattedExpensesTotal}</h1>
//         </div>
//     );
// }



