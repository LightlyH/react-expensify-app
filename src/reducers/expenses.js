// Expenses Reducer
const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [
                ...state, // ES6 spread operator
                action.expense
            ];
            // return state.concat(action.expense); // also works
        case 'REMOVE_EXPENSE':
            // return state.filter(expense => expense.id !== action.id); // also works
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_EXPENSE':
            return state.map(expense => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    };
                } else {
                    return expense;
                }
            });
        case 'SET_EXPENSES':
            return action.expenses; // don't care about previous expenses 
        
        default:
            return state;
    }
};