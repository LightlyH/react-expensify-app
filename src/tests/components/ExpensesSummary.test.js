import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from './../fixtures/expenses';

test('should correctly render ExpensesSummary with one expense', () => {
    const wrapper = shallow(<ExpensesSummary expenses={[expenses[0]]}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should correctly render ExpensesSumary with multiple expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenses={expenses}/>);
    expect(wrapper).toMatchSnapshot();
});

// this alternative also works
// test('should correctly render ExpensesSummary with one expense', () => {
//     const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={245} />);
//     expect(wrapper).toMatchSnapshot();
// });

// test('should correctly render ExpensesSumary with multiple expenses', () => {
//     const wrapper = shallow(<ExpensesSummary expenseCount={23} expensesTotal={24500}/>);
//     expect(wrapper).toMatchSnapshot();
// });