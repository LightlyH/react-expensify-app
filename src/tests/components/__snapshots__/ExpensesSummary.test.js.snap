// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSumary with multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    3
     
    expenses
     totaling 
    $0.00
  </h1>
</div>
`;

exports[`should correctly render ExpensesSummary with one expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totaling 
    $0.00
  </h1>
</div>
`;
